packages/affiliates/libs/affiliates.js

```diff
@@ -0,0 +1,192 @@
+    const blocks = await queries.getBlocks(start, end, ["timestamp", "number"]);
+    return blocks.map(block => {
+      return {
+        ...block,
+        timestamp: moment(block.timestamp.value).valueOf()
+      };
+    });
+  }
+
+  // returns array of tuples [emp address, deployer address]
+  async function getEmpDeployerHistory(address, start, end) {
+    // this query is relatively small
+    const logs = await queries.getLogsByContract(address, start, end);
+    const decode = DecodeLog(empCreatorAbi);
+    return logs.map(decode).reduce((result, log) => {
+      result.push([log.args.expiringMultiPartyAddress, log.args.deployerAddress]);
+      return result;
+    }, []);
+  }
+
+  // update this function to calculate a price.
+  // may want to instead get price based on collateral and collaterlization ratio
+  // returns a BigInt
+  function calculateValue(tokens, closestPrice, decimals = 18) {
+    return BigInt(tokens) * BigInt(parseFixed(closestPrice.toFixed(decimals), decimals).toString());
+  }
+
+  // pure function to seperate out queries from calculations
+  // this is adapted from scrips/example-contract-deployer-attributation.js
+  function calculateRewards({
+    empWhitelist = [],
+    snapshotSteps = 64,
+    totalRewards,
+    tokenPrices,
+    blocks,
+    balanceHistories,
+    empDeployers,
+    tokenDecimals
+  }) {
+    assert(tokenPrices, "requires token prices");
+    assert(blocks, "requires blocks");
+    assert(balanceHistories, "requires balanceHistories");
+    assert(empDeployers, "requires empDeployers");
+
+    balanceHistories = new Map(balanceHistories);
+    empDeployers = new Map(empDeployers);
+    const rewardsPerBlock = BigInt(totalRewards) / BigInt(blocks.length);
+    const payoutPerSnapshot = rewardsPerBlock * BigInt(snapshotSteps);
+
+    const valuePerSnapshot = blocks.reduce((result, block, index) => {
+      if (index % snapshotSteps !== 0) return result;
+      const { timestamp } = block;
+
+      const valueByEmp = empWhitelist.reduce((result, empAddress, empIndex) => {
+        try {
+          const { tokens } = balanceHistories.get(empAddress).history.lookup(block.number);
+          const decimals = tokenDecimals[empIndex];
+          assert(decimals, "requires token decimals lookup");
+          const [, closestPrice] = tokenPrices[empIndex].closest(timestamp);
+          const totalTokens = Object.values(tokens).reduce((a, b) => a + BigInt(b), 0n);
+          // console.log({closestPrice,totalTokens},BigInt(parseFixed(closestPrice.toFixed(18),18)))
+          // need to conver total tokens to consistent decimals across all tokens
+          result.push([empAddress, calculateValue(totalTokens, closestPrice, decimals)]);
+          return result;
+        } catch (err) {
+          // this error is ok, it means we have block history before the emp had
+          // any events. this essentially means value locked at emp is 0 at this block.
+          result.push([empAddress, 0]);
```

Comment:
So this calculates the total # of tokens (minted) for each EMP?

Reply:
essentially yes, you can think of it as a value that is comparable to all other emps. this gives us a way to determine the proportions for rewards.

Chris is working on a way to improve this, but as it is, the "value" is total tokens * price of the token at that block time.